src/python/pants/backend/python/tasks/build_local_python_distributions.py

```diff
@@ -262,24 +262,17 @@ def _create_dist(self,
     setup_py_snapshot_version_argv = self._generate_snapshot_bdist_wheel_argv(
       snapshot_fingerprint, is_platform_specific)
 
-    cmd = safe_shlex_join(setup_requires_pex.cmdline(setup_py_snapshot_version_argv))
+    cmd = safe_shlex_join(setup_py_runner.cmdline(setup_py_snapshot_version_argv))
     with self.context.new_workunit('setup.py', cmd=cmd, labels=[WorkUnitLabel.TOOL]) as workunit:
-      with pushd(dist_target_dir):
-        result = setup_requires_pex.run(args=setup_py_snapshot_version_argv,
-                                        stdout=workunit.output('stdout'),
-                                        stderr=workunit.output('stderr'))
-        if result != 0:
-          raise self.BuildLocalPythonDistributionsError(
-            "Installation of python distribution from target {target} into directory {into_dir} "
-            "failed (return value of run() was: {rc!r}).\n"
-            "The pex with any requirements is located at: {interpreter}.\n"
-            "The host system's compiler and linker were used.\n"
-            "The setup command was: {command}."
-            .format(target=dist_tgt,
-                    into_dir=dist_target_dir,
-                    rc=result,
-                    interpreter=setup_requires_pex.path(),
-                    command=setup_py_snapshot_version_argv))
+      try:
+        setup_py_runner.run_setup_command(source_dir=Path(dist_target_dir),
+                                          setup_command=setup_py_snapshot_version_argv,
+                                          stdout=workunit.output('stdout'),
+                                          stderr=workunit.output('stderr'))
+      except SetupPyRunner.CommandFailure as e:
+        raise self.BuildLocalPythonDistributionsError(
+          f"Installation of python distribution from target {dist_tgt} into directory "
+          f"{dist_target_dir} failed using the host system's compiler and linker: {e}")
```

Comment:
It's a best practice to always use the `repr()` for an exception to ensure that the type of the exception is included, like `ValueError`. Otherwise, the str representation sometimes will only have the actual exception message.

```suggestion
          f"{dist_target_dir} failed using the host system's compiler and linker: {e!r}")
```

Reply:
Why is that best practice? That seems like an odd leak to a user if the exception bubbles all the way out - they don't and shouldn't know class names. For a developer, they should be able to follow the back trace.

Reply:
The recommendation comes from a Pycon 2019 talk `Exceptional Exceptions`. https://youtu.be/V2fGAv2R5j8?t=548 to 10:21. 

Re-watching that one minute clip, their recommendation seems to be to _generally_ use `repr` rather than `str`. I buy that there are places you shouldn't, such as user-facing exceptions.

cc @asherf who frequently uses the `e!r` idiom as well.

Reply:
OK - finally listened to that bit of the talk. I totally don't buy his motivating cases and stick to my reasoning above. His cases all involve horrible exception messages in the 1st place. Either none at all, or ambiguous; e.g. "one" or "error". If the exception message is good you can probably find what code threw it via grep. If you have backtraces un-suppresssed, you can do so with no regard to exception message at all.

Reply:
I'll keep this in mind!